### PR TITLE
Implement the outline of a push scheduler service

### PR DESCRIPTION
**Pull request by ahal**
Fixes issue #1300

### CODE REVIEW

**Comment by ahal** [comment]
This is super WIP and not ready for review. I pushed it in case anyone wanted to provide early feedback and to ensure someone else doesn't start on this from scratch in the meantime.

**Comment by ahal** [comment]
I think it would be better if we implemented a `JobInfo` dataclass that stored all the job related state (rather than passing `func` and `*args` around all over the place). I started implementing it and then realized it was distracting me more than it should, so decided to leave it for a follow up commit.

File: http_service/bugbug_http/app.py [hunk 1]
```diff
@@ -115,13 +170,12 @@ def schedule_bug_classification(model_name, bug_ids):
     job_id_mapping = {get_mapping_key(model_name, bug_id): job_id for bug_id in bug_ids}
     redis_conn.mset(job_id_mapping)
```
Comment (marco-c) [inline http_service/bugbug_http/app.py]: This part is now in schedule_job and should be removed, right? (note: move the comment too)

File: http_service/bugbug_http/app.py [hunk 2]
```diff
@@ -85,24 +91,73 @@ class UnauthorizedError(Schema):
...
-def get_mapping_key(model_name, bug_id):
-    return f"bugbug:mapping_{model_name}_{bug_id}"
+def get_mapping_key(*args):
+    """Return a key that can be used to map `args` to a job id.
+
+    If args aren't specified, a key that maps `request.path` to a job id will
+    be used instead.
+    """
+    func = get_target_func()
+    return f"bugbug:job_id:{func.__name__}:{'_'.join(args)}"
```
Comment (marco-c) [inline http_service/bugbug_http/app.py]: What if we just use the args? Since the model name is in there, we shouldn't have problems

File: http_service/bugbug_http/app.py [hunk 3]
```diff
@@ -85,24 +91,73 @@ class UnauthorizedError(Schema):
...
+
+
+def get_result_key(*args, func=None):
+    """Return a key that can be used to store the result of a job.
+
+    If args aren't specified, a key that maps `request.path` to a job result
+    will be used instead.
+    """
+    func = func or get_target_func()
```
Comment (marco-c) [inline http_service/bugbug_http/app.py]: Same here

File: http_service/bugbug_http/app.py [hunk 4]
```diff
@@ -85,24 +91,73 @@ class UnauthorizedError(Schema):
...
+    return f"bugbug:job_result:{func.__name__}:{'_'.join(args)}"
+
+
+def schedule_job(*args, job_id=None):
```
Comment (marco-c) [inline http_service/bugbug_http/app.py]: And here I'd make `func` a parameter of `schedule_job`, so we can get rid of get_target_func and make it more explicit

File: http_service/bugbug_http/app.py [hunk 5]
```diff
@@ -401,7 +449,7 @@ def batch_prediction(model_name):
     auth = headers.get(API_TOKEN)
 
     if not auth:
-        return jsonify(UnauthorizedError().dump({}).data), 401
+        return jsonify(UnauthorizedError().dump({})), 401
```
Comment (marco-c) [inline http_service/bugbug_http/app.py]: Why this change?

File: http_service/bugbug_http/app.py [hunk 6]
```diff
@@ -452,6 +500,81 @@ def batch_prediction(model_name):
...
+    else:
+        LOGGER.info("Request with API TOKEN %r", auth)
+
+    # Support the string 'autoland' for convenience.
+    if branch == "autoland":
+        branch = "integration/autoland"
+
+    url = f"https://hg.mozilla.org/{branch}/json-automationrelevance/{rev}"
+    r = requests.get(url)
```
Comment (marco-c) [inline http_service/bugbug_http/app.py]: We might want to move this to the worker side to keep the web dyno always quickly responsive, as the request could be slow.

File: http_service/bugbug_http/models.py [hunk 7]
```diff
@@ -84,6 +80,8 @@ def preload_models():
 def classify_bug(
     model_name, bug_ids, bugzilla_token, expiration=DEFAULT_EXPIRATION_TTL
 ):
+    from bugbug_http.app import get_result_key
```
Comment (marco-c) [inline http_service/bugbug_http/models.py]: Is it a problem if we add this as a top-level import?

File: http_service/bugbug_http/models.py [hunk 8]
```diff
@@ -145,3 +143,22 @@ def classify_bug(
...
     return "OK"
+
+
+def schedule_tests(revset, expiration=DEFAULT_EXPIRATION_TTL):
+    from bugbug_http.app import get_result_key
+
+    data = {
+        "tasks": ["test-macosx1014-64/debug-gtest-1proc"],
+        "groups": ["caps/test/unit/xpcshell.ini", "dom/indexedDB/test/mochitest.ini",],
+    }
```
Comment (marco-c) [inline http_service/bugbug_http/models.py]: Not sure yet how we will handle configuring different shadow schedulers with different parameters.
I see two options:
- the configuration (confidence level) is passed to the HTTP service, and the HTTP service only returns tasks/groups with confidence higher than the one given as input;
- the HTTP service returns all tasks/groups with their associated confidence level, and the client (the shadow scheduler) does the filtering on the confidence level.

If we do the second, we're going to have to change a bit the output here (something like `["test-macosx1014-64/debug-gtest-1proc", 0.67]`).

File: http_service/tests/test_push_schedules.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+import pytest
+from rq.exceptions import NoSuchJobError
+
+from bugbug_http import app
+
+# { <job id>: <result key> }
+JOBS = {}
+
+
+class RedisMock:
```
Comment (marco-c) [inline http_service/tests/test_push_schedules.py]: There is a `fakeredis` package for mocking too, but this is good too for me.

File: http_service/bugbug_http/app.py (hunk 2 above)
Comment (marco-c) [reply http_service/bugbug_http/app.py]: > I think it would be better if we implemented a JobInfo dataclass that stored all the job related state (rather than passing func and *args around all over the place). I started implementing it and then realized it was distracting me more than it should, so decided to leave it for a follow up commit.

And yes, this would probably look better, but we can leave it for a follow-up.

File: http_service/bugbug_http/app.py (hunk 1 above)
Comment (ahal) [reply http_service/bugbug_http/app.py]: No, because `schedule_job` sets a mapping key that contains the entire list of bug ids, whereas this is setting mapping keys for each individual bug id. This is needed in case someone calls:
`/model/predict/batch` (while posting bug 123456)

and then later someone calls:
`/model/predict/123456`

This way a mapping key for the second call already exists and we don't do unnecessary work.

File: http_service/bugbug_http/app.py (hunk 2 above)
Comment (ahal) [reply http_service/bugbug_http/app.py]: It would work for now, but technically args aren't unique across jobs. The unique identifier would be function + args. Imagine we add a `/commit/<rev>/schedules` API in the future. Then a `<rev>` could be both a valid argument to the `/push` and the `/commit` endpoints. At that point, we'd need the function name to help uniquely identify those jobs. So I think it's best to leave it in.

I guess we could remove the `bugbug:job_id` prefix, but I figured it might help if we ever needed to inspect the redis store manually. I'm also not sure what other types of keys are being placed in there, so it helps avoid conflicts in general.

File: http_service/bugbug_http/app.py (hunk 4 above)
Comment (ahal) [reply http_service/bugbug_http/app.py]: Will do, I was trying to avoid needing to pass it around everywhere.. but agree it ended up being more fragile than it needed to.

File: http_service/bugbug_http/app.py (hunk 5 above)
Comment (ahal) [reply http_service/bugbug_http/app.py]: It is broken at the moment and raises `AttributeError: dict objects have no attribute 'data'`. This is because the `.data` is being called on the return value from `dump` (which is just a dict) as opposed to the return value from `jsonify` (which is a flask `Response` object).

In my testing, returning a `Response` object seems to work just as well and I thought it might be more appropriate than returning a dict here. Though maybe that's a bad assumption.

File: http_service/bugbug_http/models.py (hunk 7 above)
Comment (ahal) [reply http_service/bugbug_http/models.py]: Yeah, it's a circular import :/

This could definitely be implemented better.. maybe when we create a `JobInfo` class it can move to a utility file or something.

File: http_service/bugbug_http/models.py (hunk 8 above)
Comment (ahal) [reply http_service/bugbug_http/models.py]: I think I like the second approach better. Though it will be easy to adjust on the client side so I think it's fine to keep it simple for now and adjust as needed when we run into it.

File: http_service/tests/test_push_schedules.py (hunk 9 above)
Comment (ahal) [reply http_service/tests/test_push_schedules.py]: It's simple enough and avoids another dependency :). Feel free to fix in a follow-up if you like.

File: http_service/bugbug_http/app.py (hunk 1 above)
Comment (marco-c) [reply http_service/bugbug_http/app.py]: Oh, OK. I wonder if that could be a problem, is there a limit on the length of the key?

File: http_service/bugbug_http/models.py (hunk 7 above)
Comment (marco-c) [reply http_service/bugbug_http/models.py]: OK, let's file a follow-up so we don't forget

File: http_service/bugbug_http/app.py (hunk 1 above)
Comment (ahal) [reply http_service/bugbug_http/app.py]: Apparently the redis key size limit is 512MB.. so we're probably good :)

That said, we don't actually use the key that has all the bug ids in it.. we only use that to schedule the job, and then the worker adds results for each individual bug. So we could potentially not set that key if you are worried (though doing so makes the two endpoints diverge more in implementation).

**Comment by ahal** [comment]
Please make sure the bug classification endpoint is working as expected before deploying live! I wasn't sure how to test it, and the `http_service/tests` aren't very thorough.
